Add a blog post covering Daniel's work on Spyder 6 and beyond (#38)
spyder-ide/spyder-website: src/routes/blog/spyder-6-daniel/+page.md

---
title: "Spyder 6 under the hood: Editor migration, remote dev QA, test overhaul and more!"
author:
  - dalthviz

>   - camgerlach

Reviewer: Shouldn't @ccordoba12 be included here? Also, I think it makes sense to put @CAM-Gerlach as first author. I see the blog post as kind of like a chronicle where CAM compiled what I have been doing around, reasons of that work prioritization and how the work was done

Author: Okay sure 👍 

```suggestion
author:
  - camgerlach
  - dalthviz
  - ccordoba12
```

tags: Feature, Spyder6
category: Releases
pub_date: 2024-12-13

> summary: "Beyond the headline features, there's a lot more new and improved under the hood in Spyder 6. Daniel Althviz, Spyder's release manager and co-maintainer, was at the forefront of much of it, and we're here to share the highlights with all of you and what he plans to work on next!"

Reviewer 2: I really like this summary, it's pretty cool!

---

[Daniel Althviz](https://github.com/dalthviz) is one of our longest-tenured Spyder developers, having started working on the project near the end of 2016.
Now eight years later, he's the project's co-maintainer and its our release manager, responsible for the many steps required to release Spyder across all our various platforms.

> On top of that, he is also the lead maintainer of two of the most important libraries on which Spyder, and a lot of other software depends: the [QtPy](https://github.com/spyder-ide/qtpy) GUI abstraction layer and the [QtAwesome](https://github.com/spyder-ide/qtawesome) icon font framework.

Reviewer 2: ```suggestion
On top of that, he is also the lead maintainer of two of the most important libraries on which Spyder, and a lot of other scientific projects depend on: the [QtPy](https://github.com/spyder-ide/qtpy) GUI abstraction layer and the [QtAwesome](https://github.com/spyder-ide/qtawesome) icon font framework.
```

* Make emphasis on scientific projects because that's our core audience.
* I think `on` is missing? Not sure about it though.

Author: ```suggestion
On top of that, he is also the lead maintainer of two of the most important libraries on which Spyder, and a lot of other scientific software, depends: the [QtPy](https://github.com/spyder-ide/qtpy) GUI abstraction layer and the [QtAwesome](https://github.com/spyder-ide/qtawesome) icon font framework.
```

> In 2023, Spyder received a [Chan Zuckerberg Initiative](https://chanzuckerberg.com/) Essential Open Source Software for Science [Cycle 5 grant, number 384](https://chanzuckerberg.com/eoss/proposals/enhancing-spyder-ide-remote-support-for-scientific-research-in-python/) to support a new remote development architecture and features in Spyder and related development.

Reviewer 2: ```suggestion
In 2023, Spyder received a [Chan Zuckerberg Initiative](https://chanzuckerberg.com/) Essential Open Source Software for Science [Cycle 5 grant](https://chanzuckerberg.com/eoss/proposals/enhancing-spyder-ide-remote-support-for-scientific-research-in-python/) (number 384) to support a new remote development architecture and features in Spyder and related development.
```

De-emphasize the number a bit because it's not critical for most readers.

Author: ```suggestion
In 2023, Spyder received a [Chan Zuckerberg Initiative](https://chanzuckerberg.com/) Essential Open Source Software for Science Cycle 5 grant, [number 384](https://chanzuckerberg.com/eoss/proposals/enhancing-spyder-ide-remote-support-for-scientific-research-in-python/) to support a new remote development architecture and features in Spyder and related development.
```

This has helped us fund Daniel's work on many of his contributions to Spyder 6 connected to that!

> Thanks to all of that, Daniel has put in a lot of work toward Spyder 6's development.

Reviewer 2: ```suggestion
Thanks to that, Daniel has put in a lot of work toward Spyder 6's development.
```

I don't understand the meaning of `all` here, since the previous paragraph is talking about a single thing: the CZI grant. So, I'd suggest to remove it.

Author: ```suggestion
Thanks to all his experience and responsibilities, Daniel has put in a lot of work toward Spyder 6's development.
```

Read on to hear the highlights!<br><br>

![The Spyder Contributions page on GitHub, showing a graph of all contributions over time, and Daniel highlighted as the number two contributor between Feb 1 2023 and the present, after lead maintainer Carlos Cordoba](daniel-contributions.png "Daniel has the second-most contributions to the core Spyder repository since Feb 2023")

## Spyder Editor migration to new plugin API

We introduced a brand new plugin API with Spyder 5, and as part of that refactored most of Spyder's panes and many other of its internal components to be plugins that use it, rather than tightly coupled with Spyder itself.

> That both makes Spyder much more modular and easier to maintain, but also gives external plugins all the same power and flexibility that those Spyder internal panes and components have!

Reviewer 2: ```suggestion
That both makes Spyder much more modular and easier to maintain, but also gives external plugins all the same power and flexibility that those internal panes and components have!
```

It sounds repetitive to me.

Author: ```suggestion
That both makes Spyder much more modular and easier to maintain, and gives external plugins all the same power and flexibility that internal panes and components have!
```

> However, the one major exception to that in Spyder 5 was the Editor, which wasn't migrated yet due to the high level of complexity and the further new features that would need to be added to the plugin API.

Reviewer 2: ```suggestion
However, the one major exception to that in Spyder 5 was the Editor, which hadn't been migrated due to the high level of complexity and the further new features that would need to be added to the plugin API.
```

It's easier to read like this for me as a non-native speaker.

> Migrating the Editor to use the modular plugin API was the final piece of the puzzle to complete the API migration making Spyder a fully modular, pluggable application, avoid having to maintain two duplicate APIs and allow interoperability with other plugins, such as those being developed for remote development in Spyder 6.

Reviewer 2: ```suggestion
Migrating the Editor to the modular API was the final piece of the puzzle to make Spyder a fully pluggable application, avoid having to maintain two duplicate APIs and enhance interoperability with other plugins, such as those being developed for remote development in Spyder 6.
```

Decrease verbosity and make the text flow better.

> Therefore, along with its related API additions and improvements, it also was a prerequisite for fully implementing our new remote development architecture.

Reviewer 2: ```suggestion
Specifically, it was a prerequisite for implementing fully remote editing capabilities, which will arrive in Spyder 6.2.
```

* This sentence was kind of repeating what was mentioned in the previous one. So, I clarified the importance of this work.
* Change `therefore` for `specifically` because the next sentence also starts with with it.

Author: ```suggestion
Specifically, it was a prerequisite for implementing remote editing capabilities, which will arrive in Spyder 6.2.
```

> Therefore, we prioritized it for Spyder 6, with Daniel taking the lead on implementing it!

Reviewer: I think this sounds kind of like nothing was done up to the point I got in charge of the Editor migration work. However, I would say that a lot of work was done for people like Edgar and Stephanie and even Gonzalo. I would say here something like:

```suggestion
Therefore, we prioritized it for Spyder 6, with Daniel pushing forward a full migrated Editor implementation!
```

While also adding to the context presented before some of the work done that lead to be able to do the work I did for the Editor migration like:

* https://github.com/spyder-ide/spyder/pull/13440
* https://github.com/spyder-ide/spyder/pull/13436
* https://github.com/spyder-ide/spyder/pull/17467

Author: After discussing this with @ccordoba12 , we thought it best to credit the people involved by name instead of linking to PRs that most people probably won't check to see who did them, like this:

```suggestion
Therefore, we prioritized finishing this migration for Spyder 6, with Daniel completing the work originally started by [Gonzalo Peña-Castellanos](https://github.com/goanpeca), [Edgar Margffoy](https://github.com/andfoy) and [Stephannie Jimenez](https://github.com/steff456).
```

Given the size of the task, Daniel split it into a number of steps, each corresponding to a GitHub Pull Request.
First, [he split](https://github.com/spyder-ide/spyder/pull/20893) the ``CodeEditor``, the main widget for editing code, into a component in charge of code completion and linting, and another one in charge of editing.
This helped separate those concerns and improve improve overall modularity and maintainability.

> Then, [he refactored the ``EditorStack`` code](https://github.com/spyder-ide/spyder/pull/21194), the widget in charge of displaying multiple tabbed ``CodeEditor`` views, to facilitate the migration.

Reviewer 2: ```suggestion
Then, [he refactored](https://github.com/spyder-ide/spyder/pull/21194) the ``EditorStack`` code, the widget in charge of displaying multiple tabbed ``CodeEditor`` views, to facilitate the migration.
```

Decrease link length.

Finally, he migrated the Editor itself to the new API in two steps: an [initial migration](https://github.com/spyder-ide/spyder/pull/21353), and a [followup cleanup](https://github.com/spyder-ide/spyder/pull/22005).
This last step alone took six months of hard work, but it will help to address a lot of technical debt accumulated throughout the years.<br><br>

![The Spyder remote kernel status dialog showing a successful connection, next to the IPython Console showing a kernel error](remote-kernel-issue-windows.png "Daniel helped find and solve remote development issues like this one")

## Remote development architecture quality testing

> Once the Editor migration was moving along, helped review, test and QA the new remote development architecture.

Reviewer 2: ```suggestion
Once the Editor migration was moving along, Daniel helped review, test and QA the new remote development architecture.
```

Seems missing.

Author: Applied in my other image moving suggestion

> This last step alone took six months of hard work, but it will help to address a lot of technical debt accumulated throughout the years.<br><br>
> 
> 
> ![The Spyder remote kernel status dialog showing a successful connection, next to the IPython Console showing a kernel error](remote-kernel-issue-windows.png "Daniel helped find and solve remote development issues like this one")
> 
> ## Remote development architecture quality testing
> 
> Once the Editor migration was moving along, helped review, test and QA the new remote development architecture.

Author: ```suggestion
This last step alone took six months of hard work, but it will help to address a lot of technical debt accumulated throughout the years.

## Remote development architecture quality testing

![The Spyder remote kernel status dialog showing a successful connection, next to the IPython Console showing a kernel error](remote-kernel-issue-windows.png "Daniel helped find and solve remote development issues like this one")

Once the Editor migration was moving along, Daniel helped review, test and QA the new remote development architecture.
```

As part of that, he also developed a number of fixes and improvements to address the issues he found.

> From that, he [implemented UI/UX improvements](https://github.com/spyder-ide/spyder/pull/22303) for the graphical components used to create a remote connection, based on the issues he found.

Reviewer 2: ```suggestion
For instance, he implemented several [UI/UX improvements](https://github.com/spyder-ide/spyder/pull/22303) for the graphical components used to create a remote connection, based on the issues he found.
```

* `From that` doesn't sound good to me. I'm not happy with `For instance` either, but I can't think of a better suggestion.
* Decrease link length and add `several` to make the sentence flow better.

Author: ```suggestion
For instance, he implemented [several UI/UX improvements](https://github.com/spyder-ide/spyder/pull/22303) for the graphical components used to create a remote connection, based on the issues he found.
```

> Furthermore, during his checks of remote connections were working on Windows, he discovered several errors with when running async code, which [he fixed](https://github.com/spyder-ide/spyder/pull/22137).

Reviewer 2: ```suggestion
Furthermore, during his checks of remote connections working on Windows, he discovered several errors with running async code, which he helped [to fix](https://github.com/spyder-ide/spyder/pull/22137).
```

* `were` and `with` seem unnecessary to me (I can't understand the meaning of this sentence otherwise).
* Hendrick fixed those issues but Daniel helped a lot with testing.

Reviewer: Was going to say the same. The one that actually fixed stuff was Hendrick. I just pointed out issues and helped check if Henrick's proposed fixes actually fixed the issues I encountered.

Author: ```suggestion
Furthermore, during his checks of remote connections working on Windows, he discovered several errors with running async code, which he helped [fix](https://github.com/spyder-ide/spyder/pull/22137).
```

> Finally, he also encountered problems when trying to tunnel ports through SSH, which he [also solved](https://github.com/spyder-ide/spyder/pull/22223](https://github.com/spyder-ide/spyder/pull/22223).

Reviewer 2: ```suggestion
Finally, he also encountered problems when trying to tunnel ports through SSH, which he also contributed [to solve](https://github.com/spyder-ide/spyder/pull/22223).
```

* Clarify Daniel's role.
* Fix link syntax.

Author: ```suggestion
Finally, he also encountered problems when trying to tunnel ports through SSH, which he also contributed to [solving](https://github.com/spyder-ide/spyder/pull/22223).
```

> This all helped make our remote development architecture much more stable and usable.

Reviewer: Probably here a link to the more detailed blog post about the remote development implementation from HEndrick could makes sense. So something like:

```suggestion
This all helped make our remote development architecture much more stable and usable. If you want to know more about the remote development work and Henrick's experience while working on it give a check to [his blog post!](https://www.spyder-ide.org/blog/spyder-6-remote-development/)
```

Author: ```suggestion
This all helped make our remote development architecture much more stable and usable.
To learn more about Spyder's new remote development architecture, check out [Hendrik's blog post!](../spyder-6-remote-development/).
```

- [x] Gave check
- [x] Fixed grammar
- [x] Made phrasing more idiomatic and concise
- [x] All done

## Spyder test suite overhaul

> To help support testing of Spyder 6 and its new components, Daniel helped make the Spyder test suite faster and more robust.

Reviewer 2: ```suggestion
To enhance testing of Spyder 6 and its new components, Daniel helped make the Spyder test suite faster and more robust.
```

* `helped` is used in the second sentence.
* I think the sentence sounds better this way.

The core team was witnessing constant errors and delays when running automated tests in continuous integration systems, which made it necessary for Daniel invest several weeks to fix that so that the Spyder 6 development process was smoother.

> Therefore, [Daniel's work](https://github.com/spyder-ide/spyder/pull/22077) improved this situation, as well as making the tests more efficient overall.<br><br>
> 
> 
> ![Spyder's file switcher dialog, with a path box to enter a file name above a list of matching files and their properties](file-switcher-improvements.png "Spyder's File Switcher, with the improvements from Daniel and Angela's work")
> 
> ## Review work and project management
> 
> Much of Daniel's responsibilities involve reviewing, mentoring and assisting with the work of other developers, an important role in any project.

Author: ```suggestion
Therefore, [Daniel's work](https://github.com/spyder-ide/spyder/pull/22077) improved this situation, as well as making the tests more efficient overall.

## Review work and project management

![Spyder's file switcher dialog, with a path box to enter a file name above a list of matching files and their properties](file-switcher-improvements.png "Spyder's File Switcher, with the improvements from Daniel and Angela's work")

Much of Daniel's responsibilities involve reviewing, mentoring and assisting with the work of other developers, an important role in any project.
```

> Daniel helped [Ryan Clary](https://github.com/mrclary), the developer who created a new set of installers for Spyder 6 using [Conda-Constructor](https://conda.github.io/constructor/), which are more powerful, flexible and updateble, and support installing packages, environments and Spyder plugins locally and remotely, as needed for the new remote development functionality among others.
> In addition, he tested and validated the installers' update process, as well as finding and reporting several bugs on Windows.

Reviewer 2: ```suggestion
Daniel helped [Ryan Clary](https://github.com/mrclary), the developer who created a new [set of installers](../cbi-installers) for Spyder 6, test and validate the installers' update process, as well as finding and reporting several bugs on Windows during the year-long process that took Ryan to stabilize them.
```

* Remove unnecessary sentence about the installers to emphasize Daniel's work.
* Mention how long it took Ryan to stabilize the installers, which makes Daniel's work more relevant.

Author: ```suggestion
Daniel helped [Ryan Clary](https://github.com/mrclary), the developer who created our new [Conda-based installers](../cbi-installers) for Spyder 6, test and validate the installers and their update process for each release in the year-long process of stabilizing them, as well as finding and reporting many bugs on Windows.
```

Daniel also reviewed and assisted with [Angela Remolina](https://github.com/AngelaRemolina)'s work on major improvements to Spyder's File/Symbol Switcher.

> Angela, an intern who worked for on Spyder over several months, migrated the File/symbol switcher to a plugin, so it is modular and accessible to other plugins.

Reviewer 2: ```suggestion
Angela, an intern who worked on Spyder over several months, migrated the File/Symbol switcher to a plugin, so it is modular and accessible to other plugins.
```

* `for` seems unnecessary.
* `File/Symbol` are both capitalized in the previous sentence.

Author: Fixed a confirmed typo in your comment

> As part of her project he helped manage, she also added lightning-fast, [ripgrep](https://github.com/BurntSushi/ripgrep)-accelerated searching of all files in the current project.

Reviewer 2: ```suggestion
As part of this project, she also added the initial support for lightning-fast, [fzf](https://github.com/junegunn/fzf)-accelerated searching of all the current project files in the switcher.
```

* The initial part of this sentence is very convoluted, so I simplified it.
* Angela only added the initial support to search for files (I had to finish that).
* We're using `fzf` for that, not `ripgrep`.
* Make a connection with her previous switcher work.

Author: ```suggestion
As part of this project, she also added the initial support in the switcher for lightning-fast, [fzf](https://github.com/junegunn/fzf)-accelerated searching of all files in the current project.
```

> Finally, along with [Juan-Sebastian Bautista](https://github.com/jsbautista) whom he has mentored and managed, Daniel performed an in-depth review of the numerous UI/UX enhancements included in Spyder 6.

Reviewer 2: ```suggestion
Finally, Daniel performed an in-depth review of the numerous UI/UX enhancements that [Juan-Sebastian Bautista](https://github.com/jsbautista) worked on for Spyder 6, whom Daniel has mentored and managed since he joined the project two years ago.
```

In its current wording, this sentence means to me that Juan and Daniel reviewed the work of others, which is not correct. So, I rephrased it to make it clearer.

Author: ```suggestion
Finally, Daniel performed an in-depth review of the numerous UI/UX enhancements for Spyder 6 developed by [Juan-Sebastian Bautista](https://github.com/jsbautista), whom Daniel has mentored and managed for the past two years.
```

> This work has made Spyder much easier and more intuitive to use, including for users with accessibility needs and with translated versions.

Reviewer 2: ```suggestion
This work has made Spyder much easier and more intuitive to use, including for users with accessibility needs.
```

I really don't understand what you mean by the `translated versions` part. So, I'd propose to remove it.

Thanks to Daniel's mentorship, Juan-Sebastian looks to continue his work with Spyder into the future.<br><br>

![Spyder release page, showing a list of version numbers and their corresponding information and files](daniel-release-page.png "Daniel has released dozens of recent Spyder versions")

## Spyder release management

> Since February 1, 2023, thanks to the CZI EOSS 5 grant funding, Daniel has also been active as Spyder's Release Manager, taking on the substantial task of handling the many critical steps required to produce stable packages for the many platforms we support, including:

Reviewer 2: ```suggestion
Since February 1, 2023, thanks to the CZI grant funding mentioned before, Daniel has also been active as Spyder's Release Manager, taking on the substantial task of handling the many critical steps required to produce stable packages for the many platforms we support, including:
```

Remove unnecessary details and make the text flow better.

Author: Applied in the other suggestion

> Thanks to Daniel's mentorship, Juan-Sebastian looks to continue his work with Spyder into the future.<br><br>
> 
> 
> ![Spyder release page, showing a list of version numbers and their corresponding information and files](daniel-release-page.png "Daniel has released dozens of recent Spyder versions")
> 
> ## Spyder release management
> 
> Since February 1, 2023, thanks to the CZI EOSS 5 grant funding, Daniel has also been active as Spyder's Release Manager, taking on the substantial task of handling the many critical steps required to produce stable packages for the many platforms we support, including:

Author: ```suggestion
Thanks to Daniel's mentorship, Juan-Sebastian looks to continue his work with Spyder into the future.

## Spyder release management

![Spyder release page, showing a list of version numbers and their corresponding information and files](daniel-release-page.png "Daniel has released dozens of recent Spyder versions")

Since February 1, 2023, thanks to the CZI grant funding mentioned before, Daniel has also been active as Spyder's Release Manager, taking on the substantial task of handling the many critical steps required to produce stable packages for the many platforms we support, including:
```

* [PyPI](https://pypi.org/project/spyder/) (pip)
* [Conda-Forge](https://anaconda.org/conda-forge/spyder) (Conda)
* [Spyder installers](https://www.spyder-ide.org/download/), and the
* [GitHub releases](https://github.com/spyder-ide/spyder/releases).

Since then, he has served as Release Manager for one minor and eight bugfix releases for Spyder 5 (from 5.4.3 to 5.5.6), and eight pre-releases, one major release and three bugfix releases for Spyder 6 (from 6.0 alpha2 to 6.0.3).

> Without his work, our users wouldn't have been able to benefit from all the testing and stable releases that we have, especially for Spyder 6!

Reviewer 2: ```suggestion
Without his work, our users wouldn't have been able to benefit from all the testing and stable releases that we've had, especially for Spyder 6!
```

Seems more correct to me.

Author: ```suggestion
Without his work, our users wouldn't have been able to benefit from all the developmental and stable releases that we've had, especially for Spyder 6!
```

## What's next?

> In the immediate future, Daniel will continue his critical role as the Release Manager for Spyder, ensuring users continue to see regular stable updates for Spyder 6.

Reviewer 2: ```suggestion
In the immediate future, Daniel will continue his critical role as the Release Manager for Spyder, ensuring users keep seeing regular updates for Spyder 6.
```

* Replace `continue` by `keep` because `continue` is used before.
* `stable` seems redundant to me.

> Additionally, he'll further his partnership with Juan-Sebastian, this time focused on maintaining Spyder 6 and fixing any further bugs that are found.

Reviewer 2: ```suggestion
Additionally, he'll further his partnership with Juan-Sebastian, this time focused on maintaining Spyder 6.0 and fixing any further bugs that are found.
```

> He'll also work on further improving [Jupyter QtConsole](https://qtconsole.readthedocs.io/en/stable/), the frontend that Spyder (in the IPython Console) and other users and applications employ to provide a rich, interactive graphical interface to local and remote kernels.

Reviewer 2: ```suggestion
He'll also work on improving [Jupyter QtConsole](https://qtconsole.readthedocs.io/en/stable/), the frontend that Spyder (in the IPython Console) and other users and applications employ to provide a rich, interactive graphical interface to local and remote kernels.
```

`further` was used in the previous sentence.

> Finally, he'll help implement the frontend interface to remote file systems added as part of our remote development work, allowing users to manipulate remote files just like those locally!

Reviewer 2: ```suggestion
Finally, he'll help implement the frontend interface to remote file systems, as part of our remote development work, allowing users to manipulate remote files and directories just like those locally!
```

Make sentence flow better.

Author: ```suggestion
Finally, he'll help implement the frontend interface to remote file systems as part of our remote development work, allowing users to manipulate remote files and directories just like those locally!
```

---
title: "Spyder 6 under the hood: Editor migration, remote dev QA, test overhaul and more!"
author:
  - camgerlach
  - dalthviz
  - ccordoba12
tags: Feature, Spyder6
category: Releases
pub_date: 2024-12-13
summary: "Beyond the headline features, there's a lot more new and improved under the hood in Spyder 6. Daniel Althviz, Spyder's release manager and co-maintainer, was at the forefront of much of it, and we're here to share the highlights with all of you and what he plans to work on next!"
---

[Daniel Althviz](https://github.com/dalthviz) is one of our longest-tenured Spyder developers, having started working on the project near the end of 2016.
Now eight years later, he's the project's co-maintainer and its our release manager, responsible for the many steps required to release Spyder across all our various platforms.
On top of that, he is also the lead maintainer of two of the most important libraries on which Spyder, and a lot of other scientific software, depends: the [QtPy](https://github.com/spyder-ide/qtpy) GUI abstraction layer and the [QtAwesome](https://github.com/spyder-ide/qtawesome) icon font framework.

> In 2023, Spyder received a [Chan Zuckerberg Initiative](https://chanzuckerberg.com/) Essential Open Source Software for Science Cycle 5 grant, [number 384](https://chanzuckerberg.com/eoss/proposals/enhancing-spyder-ide-remote-support-for-scientific-research-in-python/) to support a new remote development architecture and features in Spyder and related development.

Reviewer: I think here the parentheses from @ccordoba12 suggestion are missing?
```suggestion
In 2023, Spyder received a [Chan Zuckerberg Initiative](https://chanzuckerberg.com/) Essential Open Source Software for Science Cycle 5 grant ([number 384](https://chanzuckerberg.com/eoss/proposals/enhancing-spyder-ide-remote-support-for-scientific-research-in-python/)) to support a new remote development architecture and features in Spyder and related development.
```

Reviewer: Also, using the actual proposal name (`Enhancing Spyder IDE Remote Support for Scientific Research in Python`) than its number could make more sense?

Author: Hey, we actually went through all the suggestions in our meeting together today and adjusted the text per my applied suggestion, since the grant number is important to include per standard attribution convention and the link refers to our specific grant, that number, as opposed to cycle 5 in general. Thanks for making sure!
